feat(tactic/compute_degree + test/compute_degree): a tactic for computing `nat_degree`s of polynomials

adomani opened the pull request:
`compute_degree` tries to solve a goal of the form `f.nat_degree = d` or `f.degree = d`, where `d : ℕ` and `f`
satisfies:
* `f` is a sum of expression of the form
  `C a * X ^ n, C a * X, C a, X ^ n, X, monomial n a, monomial n a * monomial m b`,
* all exponents and the `n` in `monomial n a` are *closed* terms of type `ℕ`,
* the term with largest exponent is `X ^ n` and is the unique term of its degree (repetitions are
  allowed in terms of smaller degree).
If the given degree does not match what the tactic computes, the tactic suggests the degree it
computed.
The tactic also reports when it is used with non-closed natural numbers as exponents.

There is a further tactic `compute_degree_le`, closing goals of the form  `f.nat_degree ≤ d`, with similar, but slightly less stringent restrictions in `f`.  Currently, `compute_degree_le` does not support `degree`, only `nat_degree`.

[Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Computing.20nat_degrees.20of.20polynomials)

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

- [x] depends on: #14762 for the implementation of `compute_degree_le`
- [x] depends on: #13483 for the implementation of `move_add`

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -0,0 +1,280 @@
+variables {R : Type*} [semiring R]
+
+lemma nat_degree_monomial_le (a : R) {m n : ℕ} (mn : m ≤ n) :
+  (monomial m a).nat_degree ≤ n :=
```
This should probably be
```suggestion
lemma nat_degree_monomial_le (a : R) {m : ℕ} :
  (monomial m a).nat_degree ≤ m :=
```

adomani replied:
Your suggested change is shaper indeed.  However, I only use this lemma in the current PR and every time I use it, your version would be used as `(nat_degree_monomial_le _ _).trans_le _`.  For this reason, I preferred the current version.

Would you still want me to change this?

adomani replied:
I came around to this and the sharper formulation is the one that I PR'ed to mathlib.  I will live with `...trans_le _`!

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -0,0 +1,292 @@
+  | `(@polynomial.X %%R %%inst) := do
+      refine ``(polynomial.nat_degree_C_mul_X _ _),
+      assumption <|> exact ``(one_ne_zero)
+  | `(@has_pow.pow (@polynomial %%R %%nin) %%N %%inst %%mX %%n) := do
```
I would guess you only want to match powers by `nat`, not by a general `%%N`

adomani replied:
Yes, although the tactic will fail elsewhere if the exponent is not a closed term of type `nat` anyway: after I gather the exponents, I sort them and that process fails for non-nat terms.

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -0,0 +1,292 @@
+  | `(coe_fn $ polynomial.C) := to_expr ``(0)
+  | a := do e.get_app_args.nth 4 >>= return <|>
+         do val ← to_expr ``(polynomial.nat_degree),
+           return $ expr.mk_app val [e]
```
What case is `e.get_app_args.nth 4` trying to match?

adomani replied:
When all the previous matches fail, the process returns `e.nat_degree` for all remaining terms.  This allows `compute_degree_le` (not `compute_degree`) to work with some degenerate cases, where you are trying to compute `(X + 1).nat_degree \le 1`, where `1` would not be matched, since it is not `C 1`.

adomani replied:
In fact, dealing better with "numerals" in polynomial rings that are not obtained as `C (numeral)` is something that I struggled with a little.

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -0,0 +1,292 @@
+  | a := do e.get_app_args.nth 4 >>= return <|>
+         do val ← to_expr ``(polynomial.nat_degree),
+           return $ expr.mk_app val [e]
+  end
```
Perhaps tidier as
```suggestion
meta def extract_deg_single_term : expr → tactic expr
| `(polynomial.X) := to_expr ``(1)
| `(polynomial.X ^ %%n) := return n
| (expr.app `(⇑(polynomial.monomial %%n)) x) := return n
| (expr.app `(⇑polynomial.C) x) := to_expr ``(0)
| e := to_expr ``(polynomial.nat_degree %%e)
```

adomani replied:
This is great!  Thank you so much!  I struggled a lot to get the matching to work and you saw the result!

adomani replied:
I think that, with this template, I can now simplify a few of the other matches in the file.

adomani committed:
add f

adomani committed:
Merge branch 'master' into adomani_compute_degree

adomani committed:
update with PR

adomani committed:
fix namespace

adomani committed:
Add Eric's comment as a todo

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge remote-tracking branch 'origin/master' into aa_sort

adomani committed:
namespacing and some shifts

adomani committed:
more of Robert's suggestions

adomani committed:
simpler tests

adomani committed:
update following more suggestions

adomani committed:
more tests

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
Merge remote-tracking branch 'origin/master' into aa_sort

adomani committed:
move get_summands to meta.expr list_summands

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
remove two more pairs of parens

adomani committed:
use correct name list?summands

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
rename move_add_with_errors

adomani committed:
remove layered move_add

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
test with at h,

adomani committed:
spaces

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
remove unneeded open tactic

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
recurse more

adomani committed:
typos

adomani committed:
replace "trace" with "fail"

adomani committed:
update doc-strings to reflect failure

adomani committed:
tests with `success_if_fail`

adomani committed:
clearer doc-string

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
with comments

adomani committed:
without comments

adomani committed:
tests

adomani committed:
test

adomani committed:
doc-string and rename

adomani committed:
whitespace

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
tests

adomani committed:
add `list_summands`

adomani committed:
remove docstring in `tactic.congr`

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
undo change in `meta.expr` for CI cycles

adomani committed:
in the middle

adomani committed:
move_oper

adomani committed:
tests

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
clean up

adomani committed:
error messages

adomani committed:
todo

adomani committed:
lint

adomani committed:
doc-string

adomani committed:
typo

adomani committed:
unneeded errors?

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
spaces

adomani committed:
fix?

adomani committed:
better fix?

adomani committed:
finally!

adomani committed:
small improvement: nums

adomani committed:
remove trace

adomani committed:
fewer unused errors

adomani committed:
bit0 bit1 muldegree lemmas commented

adomani committed:
lots to cleanup

adomani committed:
cleanup test

adomani committed:
cleanup compute_degree

adomani committed:
Merge branch 'master' into adomani_compute_degree

adomani committed:
more cleanup

adomani committed:
remove compute_degree_old

adomani committed:
docstring

adomani committed:
again working

adomani committed:
tests

adomani committed:
doc-string

adomani committed:
test

adomani committed:
tests

adomani committed:
namespaces and cleanup

adomani committed:
comment for quiet down the test file

adomani committed:
remove unneeded `open`s

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
remove paragraph

adomani committed:
add Eric's code and a version for factors

adomani committed:
Update src/meta/expr.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
move `list_binary_operands` to tactic.core

adomani committed:
relate to `list_binary_operands`

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
give possibility of using custom error

adomani committed:
doc-string and rename

adomani committed:
better errors

adomani committed:
compute_degree takes a hint!

adomani committed:
one more test

adomani committed:
doc-string

adomani committed:
Merge branch 'master' into adomani_compute_degree

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
adjust to list_binary_summands

adomani committed:
better example

adomani committed:
factors

adomani committed:
better versions from code review

adomani committed:
Apply suggestions from code review

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
rename expr --> tactic

adomani committed:
Merge branch 'adomani_list_summands' of ssh://github.com/leanprover-community/mathlib into adomani_list_summands

adomani committed:
Typo

eric-wieser committed:
add test file

adomani committed:
one more failed test

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
use the more stable `tactic.list_binary_operands`

adomani committed:
remove potentially buggy code

adomani committed:
reduce tests

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
Update src/tactic/core.lean

Co-authored-by: Rob Lewis <Rob.y.lewis@gmail.com>

adomani committed:
lint

adomani committed:
feat(analysis/asymptotics): define `is_Theta` (#14567)

* define `f =Θ[l] g` and prove basic properties;
* add `is_O.const_smul_left`, `is_o.const_smul_left`;
* rename `is_O_const_smul_left_iff` and `is_o_const_smul_left_iff` to
  `is_O_const_smul_left` and `is_o_const_smul_left`.

adomani committed:
add tactic compute_degree_le

adomani committed:
add tests

adomani committed:
Merge branch 'master' into adomani_compute_degree_le

adomani committed:
docs

adomani committed:
messing with infer_type

adomani committed:
conservative_eval

adomani committed:
tests

adomani committed:
spacing

adomani committed:
eval_guessing

adomani committed:
fewer tests

adomani committed:
doc-string

adomani committed:
Merge branch 'adomani_compute_degree_le' into adomani_compute_degree

adomani committed:
doc

adomani committed:
wrong fix

adomani committed:
core and !

adomani committed:
Merge branch 'adomani_compute_degree_le' into adomani_compute_degree

adomani committed:
tests not working

adomani committed:
streamline

adomani committed:
typos and remove apply_instance

adomani committed:
Merge branch 'adomani_compute_degree_le' into adomani_compute_degree

adomani committed:
fix

adomani committed:
broken

adomani committed:
docs

adomani committed:
golf?

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
pre-cleanup

adomani committed:
works with coeff, but slow

adomani committed:
faulty

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'adomani_list_summands' into aa_sort

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
better example

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
cleared up

adomani committed:
cosmetic

adomani committed:
a test with forall, exists

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
tests

adomani committed:
lines

adomani committed:
better errors

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
Update src/tactic/move_add.lean

adomani committed:
doc-string

adomani committed:
minor changes

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
use alternative

adomani committed:
typo

adomani committed:
golf

adomani committed:
clean up tests

adomani committed:
Merge branch 'master' into adomani_compute_degree_le

adomani committed:
Merge branch 'master' into adomani_compute_degree_le

adomani committed:
shorten

adomani committed:
small golf

adomani committed:
extra test for failure

adomani committed:
example

adomani committed:
streamline tests

adomani committed:
Apply Eric's suggestions

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
formatting

adomani committed:
fix error messages in tests

adomani committed:
rename

adomani committed:
removed two spaces

adomani committed:
Update src/tactic/move_add.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser committed:
Eliminate the `lu` argument to `reorder_oper`.

There is no need to pass this into each level of recursion, and it makes the recursion a little clearer if we don't.

eric-wieser committed:
add a test of metavariables

eric-wieser committed:
replace is_given_op with as_given_op

eric-wieser committed:
use replace_goal to avoid creating metavariable issues

adomani committed:
Merge branch 'aa_sort' of ssh://github.com/leanprover-community/mathlib into aa_sort

adomani committed:
update doc-module

adomani committed:
add generic operation

adomani committed:
lint

adomani committed:
typo

adomani committed:
let test

adomani committed:
questions

adomani committed:
golf

adomani committed:
add refl after simp

adomani committed:
add Eric's example to the tests

adomani committed:
Merge branch 'adomani_compute_degree_le' into adomani_compute_degree

adomani committed:
small fix

adomani committed:
slighty better guess

adomani committed:
typo

adomani committed:
pp errors

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
docs

adomani committed:
Merge branch 'master' into aa_sort

adomani committed:
merged files

adomani committed:
Merge branch 'master' into adomani_compute_degree

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -3,33 +3,206 @@ Copyright (c) 2022 Damiano Testa. All rights reserved.
+| `(bit0 %%a)      := match num_to_nat a with
+  | some an        := some (bit0 an)
+  | none           := none
+  end
```
I think you can use monadic map notation here
```suggestion
| `(bit0 %%a)      := bit0 <$> num_to_nat a
```

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -3,33 +3,206 @@ Copyright (c) 2022 Damiano Testa. All rights reserved.
+| `(%%a + %%b)      := match num_to_nat a, num_to_nat b with
+  | (some an), (some bn) := some (an + bn)
+  | _, _ := none
+  end
```
```suggestion
| `(%%a + %%b)      := (+) <$> a <*> b
```
untested

eric-wieser replied:
Indeed you use this syntax below!

eric-wieser replied:
Oops, I guess I meant `(+) <$> num_to_nat a <*> num_to_nat b`

This applies to the rest of this match too.

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -3,33 +3,206 @@ Copyright (c) 2022 Damiano Testa. All rights reserved.
+match num_to_nat a with
+| (some an) := do
+  `(@polynomial %%R %%inst) ← infer_type a,
+  aexp ← to_expr ``(%%(reflect an)) tt ff,
```
```suggestion
  aexp ← to_expr ``(%%`(an)) tt ff,
```
should mean the same thing; unless for some reason it failed!

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -115,20 +319,149 @@ with `c` a coefficient of the polynomial `f` in question. -/
+meta def single_term_suggestions : tactic unit := do
+bo ← succeeds $ interactive.exact ``(polynomial.nat_degree_X_pow _),
+if bo then fail "Try this: exact polynomial.nat_degree_X_pow _"
+else do
```
Possibly a slightly nicer pattern:
```suggestion
ff ← succeeds $ interactive.exact ``(polynomial.nat_degree_X_pow _)
  | fail "Try this: exact polynomial.nat_degree_X_pow _",
```

adomani committed:
use namespace

adomani committed:
failing tests

adomani committed:
Eric's suggestions

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

adomani committed:
Merge branch 'adomani_compute_degree' of ssh://github.com/leanprover-community/mathlib into adomani_compute_degree

adomani committed:
Eric's suggestion of monadic notation

adomani committed:
eval_expr'

adomani committed:
tests with errors

adomani committed:
Eric's match ff

adomani committed:
tests for one-term-suggestions

adomani committed:
rm old code, incl extract_top_degree_terms_and_deg

adomani committed:
failing tests

adomani committed:
move a lemma to reduce the diff

adomani committed:
revert to original position

eric-wieser commented on src/tactic/compute_degree.lean:
```diff
@@ -3,33 +3,206 @@ Copyright (c) 2022 Damiano Testa. All rights reserved.
+sequence of `bit0` and `bit1` applied to `0` or `1`.  Otherwise, `num_to_nat e = none`. -/
+-- for the application, the line `| `(has_zero.zero) := some 0` step is unnecessary: the standard
+-- assumption is that the coefficient of the term of highest-looking term is non-zero.
+meta def num_to_nat : expr → option ℕ
+| `(has_zero.zero) := some 0
+| `(has_one.one)   := some 1
+| `(bit0 %%a)      := bit0 <$> num_to_nat a
+| `(bit1 %%a)      := bit1 <$> num_to_nat a
+| `(%%a + %%b)     := (+) <$> num_to_nat a <*> num_to_nat b
+| `(%%a * %%b)     := (*) <$> num_to_nat a <*> num_to_nat b
+| `(%%a ^ %%b)     := (^) <$> num_to_nat a <*> num_to_nat b
+| _ := none
```
This feels like it must overlap heavily with norm_num.  Perhaps @digama0 can help work out precisely how?

adomani committed:
Merge branch 'aa_sort' into adomani_compute_degree

adomani committed:
fix after merge of move_add

adomani committed:
extract issues

adomani committed:
remove #check

adomani committed:
formatting

adomani committed:
clean up

adomani committed:
remove guess_degree_to_nat

adomani committed:
formatting

adomani committed:
rename variable

adomani committed:
remove commented degree_single_terms

adomani committed:
remove compute_step

adomani committed:
tests

adomani committed:
possible fix of universes?

adomani committed:
introduce compute_degree_le_no_norm_num

adomani committed:
works, but slow

adomani committed:
remove no_norm_num

adomani committed:
add end

adomani committed:
works, but slow

adomani committed:
prune unneeded bits

adomani committed:
tests

adomani committed:
golf

adomani committed:
small cleanup

adomani committed:
remove @

adomani committed:
Merge branch 'master' into adomani_compute_degree
